src/commonMain/kotlin/fr/acinq/lightning/wire/LiquidityAds.kt

```diff
@@ -0,0 +1,153 @@
+            val sig = witness.sign(nodeKey)
+            return ChannelTlv.WillFund(sig, this)
+        }
+
+        fun write(out: Output) {
+            LightningCodecs.writeU16(fundingWeight, out)
+            LightningCodecs.writeU16(leaseFeeProportional, out)
+            LightningCodecs.writeU16(maxRelayFeeProportional, out)
+            LightningCodecs.writeU32(leaseFeeBase.sat.toInt(), out)
+            LightningCodecs.writeTU32(maxRelayFeeBase.msat.toInt(), out)
+        }
+
+        companion object {
+            fun read(input: Input): LeaseRates = LeaseRates(
+                fundingWeight = LightningCodecs.u16(input),
+                leaseFeeProportional = LightningCodecs.u16(input),
+                maxRelayFeeProportional = LightningCodecs.u16(input),
+                leaseFeeBase = LightningCodecs.u32(input).sat,
+                maxRelayFeeBase = LightningCodecs.tu32(input).msat,
+            )
+        }
+    }
+
+    /** Request inbound liquidity from a remote peer that supports liquidity ads. */
+    data class RequestRemoteFunding(val fundingAmount: Satoshi, val maxFee: Satoshi, val leaseStart: Int, val leaseDuration: Int) {
+        private val leaseExpiry: Int = leaseStart + leaseDuration
+        val requestFunds: ChannelTlv.RequestFunds = ChannelTlv.RequestFunds(fundingAmount, leaseExpiry, leaseDuration)
+
+        fun validateLeaseRates(
+            remoteNodeId: PublicKey,
+            channelId: ByteVector32,
+            remoteFundingPubKey: PublicKey,
+            remoteFundingAmount: Satoshi,
+            fundingFeerate: FeeratePerKw,
+            willFund: ChannelTlv.WillFund?
+        ): Either<ChannelException, Lease> {
+            return when (willFund) {
+                // If the remote peer doesn't want to provide inbound liquidity, we immediately fail the attempt.
+                // The user should retry this funding attempt without requesting inbound liquidity.
+                null -> Either.Left(MissingLiquidityAds(channelId))
+                else -> {
+                    val witness = LeaseWitness(remoteFundingPubKey, leaseExpiry, leaseDuration, willFund.leaseRates.maxRelayFeeProportional, willFund.leaseRates.maxRelayFeeBase)
+                    val fees = willFund.leaseRates.fees(fundingFeerate, fundingAmount, remoteFundingAmount)
+                    return if (!LeaseWitness.verify(remoteNodeId, willFund.sig, witness)) {
+                        Either.Left(InvalidLiquidityAdsSig(channelId))
+                    } else if (remoteFundingAmount <= 0.sat) {
+                        Either.Left(LiquidityRatesRejected(channelId))
+                    } else if (maxFee < fees) {
+                        Either.Left(LiquidityRatesRejected(channelId))
+                    } else {
+                        val leaseAmount = fundingAmount.min(remoteFundingAmount)
```

Comment:
I think we should check here that the remote amount is exactly what was requested, as users will probably reason in terms of lease_fee/mining_fee ratio.

Reply:
I changed that in https://github.com/ACINQ/lightning-kmp/pull/561/commits/c65c4cf919c3572095bc997868e8a01dbabb8ede, we now provide the rates as input and only accept the splice attempt if the rates haven't changed and the LSP adds at least the requested amount.